### PR TITLE
Adding a contract to store minimum required client version

### LINKED ISSUE #823: Blockchain SBAT reference governable smart contract with minimum client version
* Create a governable `BlockchainParams` smart contract as a repository for blockchain parameters that can be changed via governance. Follow the pattern for all of the other contracts.  Set sensible defaults via migration.

* First parameter: minimum client version.

* Get celo-blockchain to read this value via an ievmh call

* Get celo-blockchain to log a fatal message and exit(1) in the event its version does not meet this minimum. see `celo-blockchain/params/version.go`

### PR DESCRIPTION
**Pull request by mrsmkl**
### Description

This is a smart contract where the variables can be changed using governance. As a first variable, we have the minimum client version. In celo-blockchain PR, the client will check this variable and will exit if the required version on the blockchain is higher than current version.

### Tested

There is a unit test for the contract, not sure where to add a test where geth exits.

### Other changes

Changed migration numbering.

### Related issues

- Fixes #823

### CODE REVIEW

File: packages/protocol/contracts/governance/BlockchainParams.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
+pragma solidity ^0.5.3;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "../common/Initializable.sol";
+
+/**
+ * @title Contract for storing blockchain parameters that can be set by governance.
+ */
+contract BlockchainParams is Ownable, Initializable {
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/BlockchainParams.sol]: Naming nit: prefer not shortening if not necessary, `BlockchainParameters`.

File: packages/protocol/contracts/governance/BlockchainParams.sol [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  string public minimumClientVersion;
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/BlockchainParams.sol]: Question: do we want this to be a `string` (like a semver or something like that) or a `uint`? String would be more human readable and I guess we already have version strings inherited from geth. An int will be easier to compare, especially if anyone wants to use this value on-chain. @timmoreton, thoughts?

File: packages/protocol/migrations/15_blockchainparams.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { CeloContractName } from '@celo/protocol/lib/registry-utils'
+import { deploymentForCoreContract } from '@celo/protocol/lib/web3-utils'
+import { BlockchainParamsInstance } from 'types'
+import { config } from '@celo/protocol/migrationsConfig'
+
+const initializeArgs = async (_: string): Promise<any[]> => {
+  return [config.blockchainParams.minimumClientVersion]
+}
+
+module.exports = deploymentForCoreContract<BlockchainParamsInstance>(
```
Comment (m-chrzan) [inline packages/protocol/migrations/15_blockchainparams.ts]: This contract needs to be added to the list in the governance migration to transfer ownership.

File: packages/protocol/test/common/linkedlist.ts [hunk 4]
```diff
@@ -1,13 +1,8 @@
 import { assertRevert } from '@celo/protocol/lib/test-utils'
-// import BigNumber from 'bignumber.js'
 import { LinkedListTestContract, LinkedListTestInstance } from 'types'
 
 const LinkedListTest: LinkedListTestContract = artifacts.require('LinkedListTest')
 
-// @ts-ignore
-// TODO(mcortesi): Use BN
-LinkedListTest.numberFormat = 'BigNumber'
```
Comment (m-chrzan) [inline packages/protocol/test/common/linkedlist.ts:9]: Is this not needed anymore? I thought this was needed for some tests in the past.

File: packages/protocol/test/governance/blockchainparams.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+contract('BlockchainParams', () => {
+  let blockchainParamsTest: BlockchainParamsInstance
+  const version = '1.8.4'
+
+  beforeEach(async () => {
+    blockchainParamsTest = await BlockchainParamsTest.new()
+  })
+
+  describe('#setMinimumClientVersion()', () => {
```
Comment (m-chrzan) [inline packages/protocol/test/governance/blockchainparams.ts:23]: Add tests for `initialize`. For both functions, include a check for emitted events.

File: packages/protocol/contracts/governance/BlockchainParams.sol (hunk 2 above)
Comment (timmoreton) [reply packages/protocol/contracts/governance/BlockchainParams.sol]: I have a slight preference for a triple of ints that match how geth defines them:

```
VersionMajor = 1        // Major version component of the current release
VersionMinor = 9        // Minor version component of the current release
VersionPatch = 2        // Patch version component of the current release
```

we should be able to read and set them all at once.

it reduces the probability we set the value to something that is unparseable.

Comment (timmoreton) [reply packages/protocol/contracts/governance/BlockchainParams.sol]: these are golang `int`s  but I think it's fine to use something unsigned and smaller

File: packages/protocol/test/common/linkedlist.ts (hunk 4 above)
Comment (mrsmkl) [reply packages/protocol/test/common/linkedlist.ts:9]: Most tests for linked lists are in addresssortedlinkedlistwithmedian.ts

File: packages/celotool/src/e2e-tests/transfer_tests.ts [hunk 6]
```diff
@@ -436,6 +473,20 @@ describe('transfer tests', function(this: any) {
     })
   }
 
+  describe('when running a node', () => {
```
Comment (m-chrzan) [inline packages/celotool/src/e2e-tests/transfer_tests.ts]: It would be best to move this test out to a new e2e test file - this is not closely related with the transfer tests happening here. You'll have to set up a `ci_test_<your test name>.sh` script and put the tests in `src/e2e-tests/<your test name>.ts`. And finally modify `.circleci/config.yml` in the top-level to run the new tests in CI.

Hopefully the already existing tests are good enough examples for how to set this up, if not feel free to ping on slack.

File: packages/celotool/src/e2e-tests/transfer_tests.ts [hunk 7]
```diff
@@ -436,6 +473,20 @@ describe('transfer tests', function(this: any) {
...
+    it('should exit when minimum version is updated', async () => {
+      await restartGeth('full')
+      await setMinimumClientVersion(1, 8, 99)
+      await sleep(20)
+      try {
+        await setMinimumClientVersion(1, 8, 10)
+        throw new Error('expected failure')
+      } catch (_) {
+        return
```
Comment (m-chrzan) [inline packages/celotool/src/e2e-tests/transfer_tests.ts]: Won't we end up doing the same thing whether or not the call fails? I'm assuming we're expecting `setMinimumClientVersion` to throw, but we also throw if it doesn't.

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+   * @notice Initializes critical variables.
+   * @param _minimumClientVersion1 Minimum client version that can be used in the chain,
+   * major version.
+   * @param _minimumClientVersion2 Minimum client version that can be used in the chain,
+   * minor version.
+   * @param _minimumClientVersion3 Minimum client version that can be used in the chain,
+   * patch level.
+   */
+  function initialize(
+    uint256 _minimumClientVersion1,
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/BlockchainParameters.sol]: Naming: include `major`, `minor`, etc. in these names, not numbers. No need for `_`s if the names are not shadowing storage variables.

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    uint256 _minimumClientVersion2,
+    uint256 _minimumClientVersion3) external initializer
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/BlockchainParameters.sol]: Style: `) ...` and each modifier should go on a new line. (

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  {
+    _transferOwnership(msg.sender);
+    setMinimumClientVersion(
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/BlockchainParameters.sol]: This will probably fit on one line after renaming variables.

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+   * and 2 is the patch level.
+   */
+  function setMinimumClientVersion(uint256 major, uint256 minor, uint256 patch) public onlyOwner {
+    minimumClientVersion.major = major;
+    minimumClientVersion.minor = minor;
+    minimumClientVersion.patch = patch;
+    emit MinimumClientVersionSet(major, minor, patch);
+  }
+
+  function getMinimumClientVersion() public view
```
Comment (m-chrzan) [inline packages/protocol/contracts/governance/BlockchainParameters.sol]: Follow style guide for function declaration.

File: packages/protocol/lib/proxy-utils.ts [hunk 12]
```diff
@@ -2,13 +2,14 @@ import Web3 = require('web3')
...
+export async function setAndInitializeImplementation(
   web3: Web3,
   proxy: ProxyInstance,
   implementationAddress: string,
   initializerAbi: any,
   ...args: any[]
 ) {
   const callData = web3.eth.abi.encodeFunctionCall(initializerAbi, args)
-  return proxy._setAndInitializeImplementation(implementationAddress, callData as any)
+  const res = await proxy._setAndInitializeImplementation(implementationAddress, callData as any)
```
Comment (m-chrzan) [inline packages/protocol/lib/proxy-utils.ts]: Why is this change necessary?

File: packages/protocol/test/governance/blockchainparams.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        version.major,
+        version.minor,
+        version.patch
+      )
+      const versionQueried = await blockchainParameters.getMinimumClientVersion()
+      assert.equal(version.major, versionQueried[0].toNumber())
+      assert.equal(version.minor, versionQueried[1].toNumber())
+      assert.equal(version.patch, versionQueried[2].toNumber())
+    })
+    it('should emit the MinimumClientVersionSet event', async () => {
```
Comment (m-chrzan) [inline packages/protocol/test/governance/blockchainparams.ts:36]: Style: newline between `it`s.

File: packages/protocol/test/governance/blockchainparams.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+      assertContainSubset(log, {
+        event: 'MinimumClientVersionSet',
+        args: {
+          major: new BigNumber(version.major),
+          minor: new BigNumber(version.minor),
+          patch: new BigNumber(version.patch),
+        },
+      })
+    })
+    it('only owner should be able to set', async () => {
```
Comment (m-chrzan) [inline packages/protocol/test/governance/blockchainparams.ts:54]: Ditto.

File: packages/protocol/test/governance/blockchainparams.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+  describe('#initialize()', () => {
+    it('should set the variables', async () => {
+      await blockchainParameters.initialize(version.major, version.minor, version.patch)
+      const versionQueried = await blockchainParameters.getMinimumClientVersion()
+      assert.equal(version.major, versionQueried[0].toNumber())
+      assert.equal(version.minor, versionQueried[1].toNumber())
+      assert.equal(version.patch, versionQueried[2].toNumber())
+    })
+    it('should emit correct events', async () => {
```
Comment (m-chrzan) [inline packages/protocol/test/governance/blockchainparams.ts:72]: Ditto

**Review by m-chrzan — COMMENTED** [review]
Looks good. Just an ask around e2e tests and some style fixes. [Here's](https://solidity.readthedocs.io/en/v0.5.11/style-guide.html#function-declaration) the Solidity style guide to fix the function declarations.

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+ */
+contract BlockchainParameters is Ownable, Initializable {
+
+  struct ClientVersion {
+    uint256 major;
+    uint256 minor;
+    uint256 patch;
+  }
+
+  ClientVersion minimumClientVersion;
```
Comment (asaj) [inline packages/protocol/contracts/governance/BlockchainParameters.sol]: Please explicitly set the visibility here

File: packages/celotool/src/e2e-tests/exit_test.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    // contract.
+    const _web3 = new Web3('http://localhost:8545')
+    const _parameters = new _web3.eth.Contract(blockchainParametersAbi, blockchainParametersAddress)
+    const tx = _parameters.methods.setMinimumClientVersion(major, minor, patch)
+    const gas = await tx.estimateGas({ from: validatorAddress })
+    return tx.send({ from: validatorAddress, gas })
+  }
+
+  describe('when running a node', () => {
+    it('should exit when minimum version is updated', async () => {
```
Comment (asaj) [inline packages/celotool/src/e2e-tests/exit_test.ts:70]: Why do we expect it to exit when we decrease the minimum client version from 1.8.99 to 1.8.10?

**Review by asaj — COMMENTED** [review]
It looks like the 16_elect_validators migration was accidentally renamed to 51_elect_validators

**Comment by mrsmkl** [comment]
I renamed the migrations because they are the ones that have to run last.

File: packages/celotool/src/e2e-tests/exit_test.ts (hunk 17 above)
Comment (mrsmkl) [reply packages/celotool/src/e2e-tests/exit_test.ts:70]: It is just a generic call, every RPC call should fail after the node has exited. Perhaps I'll add a comment there.

**Comment by asaj** [comment]
Rather than rename it to 51, can we keep the sequential naming?

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * @notice Sets the minimum client version.
+   * @param major Major version.
+   * @param minor Minor version.
+   * @param patch Patch version.
+   * @dev For example if the version is 1.9.2, 1 is the major version, 9 is minor,
+   * and 2 is the patch level.
+   */
+  function setMinimumClientVersion(uint256 major, uint256 minor, uint256 patch) public onlyOwner {
```
Comment (asaj) [inline packages/protocol/contracts/governance/BlockchainParameters.sol:50]: Should this be external rather than public?

File: packages/protocol/contracts/governance/BlockchainParameters.sol [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    minimumClientVersion.minor = minor;
+    minimumClientVersion.patch = patch;
+    emit MinimumClientVersionSet(major, minor, patch);
+  }
+
+  /** @notice Query minimum client version.
+   * @return Returns major, minor, and patch version numbers.
+   */
+  function getMinimumClientVersion()
+    public
```
Comment (asaj) [inline packages/protocol/contracts/governance/BlockchainParameters.sol]: Ditto, external?

File: packages/protocol/lib/registry-utils.ts [hunk 20]
```diff
@@ -33,4 +34,5 @@ export const hasEntryInRegistry: string[] = [
   CeloContractName.StableToken,
   CeloContractName.Random,
   CeloContractName.Reserve,
+  CeloContractName.BlockchainParameters,
```
Comment (asaj) [inline packages/protocol/lib/registry-utils.ts]: Please order alphabetically

File: packages/protocol/lib/registry-utils.ts [hunk 21]
```diff
@@ -12,6 +12,7 @@ export enum CeloContractName {
   SortedOracles = 'SortedOracles',
   StableToken = 'StableToken',
   Validators = 'Validators',
+  BlockchainParameters = 'BlockchainParameters',
```
Comment (asaj) [inline packages/protocol/lib/registry-utils.ts]: Please order alphabetically

File: packages/protocol/migrations/16_governance.ts [hunk 22]
```diff
@@ -65,6 +65,7 @@ module.exports = deploymentForCoreContract<GovernanceInstance>(
       'SortedOracles',
       'StableToken',
       'Validators',
+      'BlockchainParameters',
```
Comment (asaj) [inline packages/protocol/migrations/16_governance.ts]: Please order alphabetically

**Review by asaj — APPROVED** [review]
Exciting!

File: packages/protocol/contracts/governance/BlockchainParameters.sol (hunk 18 above)
Comment (mrsmkl) [reply packages/protocol/contracts/governance/BlockchainParameters.sol:50]: I call it from the initializer, that's why it is public...
